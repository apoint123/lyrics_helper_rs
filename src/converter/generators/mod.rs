//! 歌词格式生成器模块

pub mod apple_music_json_generator;
pub mod ass_generator;
pub mod enhanced_lrc_generator;
pub mod krc_generator;
pub mod lqe_generator;
pub mod lrc_generator;
pub mod lyricify_lines_generator;
pub mod lys_generator;
pub mod qrc_generator;
pub mod spl_generator;
pub mod ttml_generator;
pub mod yrc_generator;
