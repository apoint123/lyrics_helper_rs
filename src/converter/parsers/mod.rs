//! 歌词解析器模块

pub mod apple_music_json_parser;
pub mod ass_parser;
pub mod enhanced_lrc_parser;
pub mod krc_parser;
pub mod lqe_parser;
pub mod lrc_parser;
pub mod lyricify_lines_parser;
pub mod lys_parser;
// pub mod musixmatch_parser;
pub mod qrc_parser;
pub mod spl_parser;
pub mod ttml_parser;
pub mod yrc_parser;
