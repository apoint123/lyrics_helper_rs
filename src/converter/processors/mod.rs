//! 歌词处理器模块

pub mod agent_recognizer;
pub mod batch_processor;
pub mod chinese_conversion_processor;
pub mod metadata_processor;
pub mod metadata_stripper;
pub mod syllable_smoothing;
