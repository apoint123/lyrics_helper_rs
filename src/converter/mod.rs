//! 歌词转换器核心模块

pub mod generators;
pub mod parsers;
pub mod processors;
pub mod types;
pub mod utils;

use std::collections::HashMap;

pub use types::{LyricFormat, LyricLine, LyricSyllable};

use crate::converter::{
    processors::{
        batch_processor, chinese_conversion_processor::ChineseConversionProcessor,
        metadata_processor::MetadataStore,
    },
    types::{
        ConversionInput, ConversionOptions, ConversionResult, ConversionTask, ConvertError,
        FullConversionResult, InputFile, ParsedSourceData,
    },
};
use tracing::{debug, warn};

// ==========================================================
//  顶级转换入口
// ==========================================================

/// 处理一个转换任务，可以是单文件转换或批量转换。
///
/// # 参数
///
/// * `task` - 一个 `ConversionTask` 枚举，描述了要执行的具体操作（单个或批量）。
/// * `options` - 一个 `ConversionOptions` 引用，包含应用于所有转换的配置选项。
///
/// # 返回
///
/// * `Result<ConversionResult, ConvertError>` - 成功时返回包含转换结果的 `ConversionResult`，
///   失败时返回具体的 `ConvertError`。
pub fn process_conversion_task(
    task: ConversionTask,
    options: &ConversionOptions,
) -> Result<ConversionResult, ConvertError> {
    match task {
        ConversionTask::Single(input) => {
            let full_result = convert_single_lyric(&input, options)?;
            Ok(ConversionResult::Single(full_result.output_lyrics))
        }
        ConversionTask::Batch(batch_input) => {
            let file_groups = batch_processor::discover_and_pair_files(&batch_input.input_dir)?;
            let (mut tasks, file_lookup) =
                batch_processor::create_batch_tasks(file_groups, batch_input.target_format);

            if tasks.is_empty() {
                warn!("在输入目录中未找到可执行的转换任务。");
                return Ok(ConversionResult::Batch(Vec::new()));
            }

            batch_processor::execute_batch_conversion(
                &mut tasks,
                &file_lookup,
                &batch_input.output_dir,
                options,
            )?;

            Ok(ConversionResult::Batch(tasks))
        }
    }
}

// ==========================================================
//  核心转换与生成逻辑
// ==========================================================

/// 将一个 `ConversionInput` 转换为指定格式的字符串。
///
/// # 参数
///
/// * `input` - 包含所有源文件信息和目标格式的 `ConversionInput`。
/// * `options` - 转换过程中的配置选项。
///
/// # 返回
///
/// * `Result<String, ConvertError>` - 成功时返回生成的目标格式字符串。
pub fn convert_single_lyric(
    input: &ConversionInput,
    options: &ConversionOptions,
) -> Result<FullConversionResult, ConvertError> {
    let mut source_data = parse_and_merge(input, options)?;
    let mut processed_data = source_data.clone();

    let mut metadata_store = MetadataStore::new();

    if let Some(overrides) = &input.user_metadata_overrides {
        for (key, values) in overrides {
            for value in values {
                let _ = metadata_store.add(key, value.clone());
            }
        }
        source_data.raw_metadata = overrides.clone();
    } else {
        for (key, values) in processed_data.raw_metadata.iter() {
            for value in values {
                if metadata_store.add(key, value.clone()).is_err() {
                    warn!(
                        "元数据键 '{}' 无法被规范化，但其值 '{}' 仍被添加。",
                        key, value
                    );
                }
            }
        }
    }
    metadata_store.deduplicate_values();

    let chinese_processor = ChineseConversionProcessor::new();
    chinese_processor.process(&mut processed_data.lines, &options.chinese_conversion);

    let output_lyrics = match input.target_format {
        LyricFormat::Lrc => generators::lrc_generator::generate_lrc(
            &processed_data.lines,
            &metadata_store,
            &options.lrc,
        ),
        LyricFormat::EnhancedLrc => generators::enhanced_lrc_generator::generate_enhanced_lrc(
            &processed_data.lines,
            &metadata_store,
            &options.lrc,
        ),
        LyricFormat::Ass => generators::ass_generator::generate_ass(
            &processed_data.lines,
            &metadata_store,
            false,
            &options.ass,
        ),
        LyricFormat::Ttml => generators::ttml_generator::generate_ttml(
            &processed_data.lines,
            &metadata_store,
            &options.ttml,
        ),
        LyricFormat::AppleMusicJson => {
            generators::apple_music_json_generator::generate_apple_music_json(
                &processed_data.lines,
                &metadata_store,
                options,
            )
        }
        LyricFormat::Qrc => {
            generators::qrc_generator::generate_qrc(&processed_data.lines, &metadata_store)
        }
        LyricFormat::Lqe => generators::lqe_generator::generate_lqe(
            &processed_data.lines,
            &metadata_store,
            &options.lqe,
        ),
        LyricFormat::Krc => {
            generators::krc_generator::generate_krc(&processed_data.lines, &metadata_store)
        }
        LyricFormat::Yrc => {
            generators::yrc_generator::generate_yrc(&processed_data.lines, &metadata_store)
        }
        LyricFormat::Lys => {
            generators::lys_generator::generate_lys(&processed_data.lines, &metadata_store)
        }
        LyricFormat::Spl => {
            generators::spl_generator::generate_spl(&processed_data.lines, &metadata_store)
        }
        LyricFormat::Lyl => generators::lyricify_lines_generator::generate_lyl(
            &processed_data.lines,
            &metadata_store,
        ),
        // LyricFormat::Musixmatch => Err(ConvertError::Internal(format!(
        //     "目前还不支持目标格式 '{:?}'",
        //     input.target_format
        // ))),
    }?;

    Ok(FullConversionResult {
        output_lyrics,
        source_data,
    })
}

/// 解析并合并一个包含主歌词、翻译和罗马音的完整输入。
///
/// # 参数
/// * `input` - 包含所有源文件信息的 `ConversionInput`。
/// * `options` - 转换过程中的配置选项，主要用于元数据处理。
///
/// # 返回
/// * `Result<ParsedSourceData, ConvertError>` - 成功时返回一个包含所有合并信息的 `ParsedSourceData`。
///
pub fn parse_and_merge(
    input: &ConversionInput,
    options: &ConversionOptions,
) -> Result<ParsedSourceData, ConvertError> {
    let mut main_parsed = parse_input_file(&input.main_lyric)?;

    let translation_parsed_data: Vec<_> = input
        .translations
        .iter()
        .map(parse_input_file_with_lang)
        .collect::<Result<_, _>>()?;

    let romanization_parsed_data: Vec<_> = input
        .romanizations
        .iter()
        .map(parse_input_file_with_lang)
        .collect::<Result<_, _>>()?;

    for (parsed, _lang) in translation_parsed_data
        .iter()
        .chain(romanization_parsed_data.iter())
    {
        main_parsed.raw_metadata.extend(parsed.raw_metadata.clone());
    }

    merge_lyric_lines(
        &mut main_parsed.lines,
        &translation_parsed_data,
        &romanization_parsed_data,
    );

    processors::metadata_stripper::strip_descriptive_metadata_lines(
        &mut main_parsed.lines,
        &options.metadata_stripper,
    );

    Ok(main_parsed)
}

// ==========================================================
//  辅助函数
// ==========================================================

/// 根据指定的格式解析单个歌词文件内容。
///
/// 这是一个底层的分派函数。
fn parse_input_file(file: &InputFile) -> Result<ParsedSourceData, ConvertError> {
    debug!("正在解析文件，格式为: {:?}", file.format);
    match file.format {
        LyricFormat::Lrc => parsers::lrc_parser::parse_lrc(&file.content),
        LyricFormat::EnhancedLrc => parsers::enhanced_lrc_parser::parse_enhanced_lrc(&file.content),
        LyricFormat::Krc => parsers::krc_parser::parse_krc(&file.content),
        LyricFormat::Ass => parsers::ass_parser::parse_ass(&file.content),
        LyricFormat::Ttml => parsers::ttml_parser::parse_ttml(&file.content, &Default::default()),
        LyricFormat::AppleMusicJson => {
            parsers::apple_music_json_parser::parse_apple_music_json(&file.content)
        }
        LyricFormat::Qrc => parsers::qrc_parser::parse_qrc(&file.content),
        LyricFormat::Yrc => parsers::yrc_parser::parse_yrc(&file.content),
        LyricFormat::Lys => parsers::lys_parser::parse_lys(&file.content),
        LyricFormat::Spl => parsers::spl_parser::parse_spl(&file.content),
        LyricFormat::Lqe => parsers::lqe_parser::parse_lqe(&file.content),
        LyricFormat::Lyl => parsers::lyricify_lines_parser::parse_lyl(&file.content),
        // LyricFormat::Musixmatch => parsers::musixmatch_parser::parse(&file.content),
    }
}

/// `parse_input_file` 的一个包装，同时返回解析数据和文件本身的语言标签。
fn parse_input_file_with_lang(
    file: &InputFile,
) -> Result<(ParsedSourceData, Option<String>), ConvertError> {
    Ok((parse_input_file(file)?, file.language.clone()))
}

/// 合并主歌词行与翻译、罗马音数据，将翻译和罗马音按时间戳插入到主歌词行中。
///
/// # 参数
/// * `main_lines` - 主歌词行的可变引用，将被插入翻译和罗马音。
/// * `translations` - 包含翻译歌词数据及其语言标签的元组切片。
/// * `romanizations` - 包含罗马音歌词数据及其语言标签的元组切片。
fn merge_lyric_lines(
    main_lines: &mut [LyricLine],
    translations: &[(ParsedSourceData, Option<String>)],
    romanizations: &[(ParsedSourceData, Option<String>)],
) {
    if translations.is_empty() && romanizations.is_empty() {
        return;
    }

    let mut translations_map: HashMap<u64, Vec<types::TranslationEntry>> = HashMap::new();
    for (trans_data, lang) in translations {
        for trans_line in &trans_data.lines {
            if let Some(text) = trans_line.line_text.clone() {
                let entry = types::TranslationEntry {
                    text,
                    lang: lang.clone(),
                };
                translations_map
                    .entry(trans_line.start_ms)
                    .or_default()
                    .push(entry);
            }
        }
    }

    let mut romanizations_map: HashMap<u64, Vec<types::RomanizationEntry>> = HashMap::new();
    for (roma_data, lang) in romanizations {
        for roma_line in &roma_data.lines {
            if let Some(text) = roma_line.line_text.clone() {
                let entry = types::RomanizationEntry {
                    text,
                    lang: lang.clone(),
                    scheme: None,
                };
                romanizations_map
                    .entry(roma_line.start_ms)
                    .or_default()
                    .push(entry);
            }
        }
    }

    for main_line in main_lines.iter_mut() {
        if let Some(trans_entries) = translations_map.get(&main_line.start_ms) {
            main_line.translations.extend_from_slice(trans_entries);
        }
        if let Some(roma_entries) = romanizations_map.get(&main_line.start_ms) {
            main_line.romanizations.extend_from_slice(roma_entries);
        }
    }
}
