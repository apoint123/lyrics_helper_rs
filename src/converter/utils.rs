//! 包含一些工具函数的模块。

use regex::Regex;
use std::collections::HashMap;
use std::sync::LazyLock;

use crate::converter::{LyricSyllable, types::LyricLine};

/// 辅助函数，用于安全地将偏移量应用到 u64 时间戳上
fn offset_timestamp(timestamp: u64, offset: i64) -> u64 {
    // 将 u64 时间戳转换为 i64 进行计算，以避免类型错误
    // 使用 .max(0) 确保结果不会是负数（时间戳不能为负）
    // 最后再安全地转换回 u64
    (timestamp as i64 + offset).max(0) as u64
}

/// 对歌词行向量应用一个时间偏移。
///
/// 此函数会就地修改传入的 `LyricLine` 向量，调整其中所有的时间戳。
///
/// # 参数
/// * `lines` - 一个可变的 `LyricLine` 切片。
/// * `offset_ms` - 要应用的偏移量（毫秒）。正数表示延迟歌词，负数表示提前歌词。
pub fn apply_offset(lines: &mut [LyricLine], offset_ms: i64) {
    if offset_ms == 0 {
        return;
    }

    for line in lines.iter_mut() {
        // 调整行本身的时间戳
        line.start_ms = offset_timestamp(line.start_ms, offset_ms);
        line.end_ms = offset_timestamp(line.end_ms, offset_ms);

        // 调整主歌词音节的时间戳
        for syl in line.main_syllables.iter_mut() {
            syl.start_ms = offset_timestamp(syl.start_ms, offset_ms);
            syl.end_ms = offset_timestamp(syl.end_ms, offset_ms);
        }

        // 如果存在，调整背景人声的时间戳
        if let Some(bg_section) = line.background_section.as_mut() {
            bg_section.start_ms = offset_timestamp(bg_section.start_ms, offset_ms);
            bg_section.end_ms = offset_timestamp(bg_section.end_ms, offset_ms);

            // 调整背景人声音节的时间戳
            for bg_syl in bg_section.syllables.iter_mut() {
                bg_syl.start_ms = offset_timestamp(bg_syl.start_ms, offset_ms);
                bg_syl.end_ms = offset_timestamp(bg_syl.end_ms, offset_ms);
            }
        }
    }
}

static METADATA_TAG_REGEX: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^\[(?P<key>[a-zA-Z]+):(?P<value>.*)\]$").expect("编译 METADATA_TAG_REGEX 失败")
});

/// 尝试将一行文本解析为 LRC 风格的 `[key:value]` 元数据。
/// 如果成功，则将结果存入 `raw_metadata` 并返回 `true`。
///
/// # 返回
/// `true` - 如果该行是有效的元数据标签并已处理。
/// `false` - 如果该行不是元数据标签。
pub fn parse_lrc_metadata_tag(line: &str, raw_metadata: &mut HashMap<String, Vec<String>>) -> bool {
    if let Some(caps) = METADATA_TAG_REGEX.captures(line)
        && let (Some(key), Some(value)) = (caps.name("key"), caps.name("value"))
    {
        raw_metadata
            .entry(key.as_str().to_string())
            .or_default()
            .push(value.as_str().to_string());
        return true;
    }
    false
}

/// 处理音节的原始文本，从中分离出空格信息和干净的文本。
///
/// # 参数
/// * `raw_text_slice` - 两个时间戳之间的原始文本切片。
/// * `syllables` - 正在构建的音节列表，此函数可能会修改其中最后一个元素（为其添加尾随空格）。
///
/// # 返回
/// * `Some((clean_text, ends_with_space))` - 如果原始文本包含有效内容。元组中第一个元素是
///   去除了所有前后空格的纯净文本，第二个元素是一个布尔值，代表这个纯净文本在原始文本中
///   是否有尾随空格。
/// * `None` - 如果原始文本只包含空格，意味着它已被作为前一个音节的尾随空格处理完毕。
pub fn process_syllable_text(
    raw_text_slice: &str,
    syllables: &mut [LyricSyllable],
) -> Option<(String, bool)> {
    let has_leading_space = raw_text_slice.starts_with(char::is_whitespace);
    let has_trailing_space = raw_text_slice.ends_with(char::is_whitespace);
    let clean_text = raw_text_slice.trim();

    if has_leading_space && let Some(last_syllable) = syllables.last_mut() {
        last_syllable.ends_with_space = true;
    }

    if !clean_text.is_empty() {
        Some((clean_text.to_string(), has_trailing_space))
    } else {
        None
    }
}

/// 规范化文本中的空白字符
pub fn normalize_text_whitespace(text: &str) -> String {
    let trimmed = text.trim();
    if trimmed.is_empty() {
        return String::new();
    }
    trimmed.split_whitespace().collect::<Vec<&str>>().join(" ")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::converter::types::LyricSyllable;

    // 处理一个正常的音节
    #[test]
    fn test_process_basic_syllable_no_spaces() {
        let mut syllables: Vec<LyricSyllable> = Vec::new();
        let raw_text = "word";

        let result = process_syllable_text(raw_text, &mut syllables);

        assert_eq!(result, Some(("word".to_string(), false)));
        assert!(syllables.is_empty(), "不应修改音节列表");
    }

    // 处理一个带前导空格的音节
    #[test]
    fn test_process_syllable_with_leading_space() {
        let mut syllables = vec![LyricSyllable {
            text: "previous".to_string(),
            ..Default::default()
        }];
        let raw_text = " word";

        let result = process_syllable_text(raw_text, &mut syllables);

        assert_eq!(result, Some(("word".to_string(), false)));
        assert_eq!(syllables.len(), 1);
        assert!(
            syllables[0].ends_with_space,
            "前一个音节应被标记为有尾随空格"
        );
    }

    // 处理一个带尾随空格的音节
    #[test]
    fn test_process_syllable_with_trailing_space() {
        let mut syllables: Vec<LyricSyllable> = Vec::new();
        let raw_text = "word ";

        let result = process_syllable_text(raw_text, &mut syllables);

        assert_eq!(result, Some(("word".to_string(), true)));
        assert!(syllables.is_empty(), "不应修改音节列表");
    }

    // 处理一个同时带前导和尾随空格的音节
    #[test]
    fn test_process_syllable_with_both_spaces() {
        let mut syllables = vec![LyricSyllable {
            text: "previous".to_string(),
            ..Default::default()
        }];
        let raw_text = " word ";

        let result = process_syllable_text(raw_text, &mut syllables);

        assert_eq!(result, Some(("word".to_string(), true)));
        assert_eq!(syllables.len(), 1);
        assert!(
            syllables[0].ends_with_space,
            "前一个音节应被标记为有尾随空格"
        );
    }

    // 处理纯空格
    #[test]
    fn test_pure_whitespace_slice_modifies_previous() {
        let mut syllables = vec![LyricSyllable {
            text: "previous".to_string(),
            ..Default::default()
        }];
        let raw_text = "   ";

        let result = process_syllable_text(raw_text, &mut syllables);

        assert_eq!(result, None, "纯空格不应返回任何新音节");
        assert_eq!(syllables.len(), 1);
        assert!(
            syllables[0].ends_with_space,
            "前一个音节应被标记为有尾随空格"
        );
    }

    // 处理空的文本
    #[test]
    fn test_empty_slice_returns_none() {
        let mut syllables: Vec<LyricSyllable> = Vec::new();
        let raw_text = "";

        let result = process_syllable_text(raw_text, &mut syllables);

        assert_eq!(result, None);
        assert!(syllables.is_empty());
    }

    // 处理前导空格，但音节列表为空
    #[test]
    fn test_leading_space_with_empty_syllables_list() {
        let mut syllables: Vec<LyricSyllable> = Vec::new();
        let raw_text = " word";

        let result = process_syllable_text(raw_text, &mut syllables);

        // 确保不会崩溃
        assert_eq!(result, Some(("word".to_string(), false)));
        assert!(syllables.is_empty());
    }

    // 处理纯空格，但音节列表为空
    #[test]
    fn test_pure_whitespace_with_empty_syllables_list() {
        let mut syllables: Vec<LyricSyllable> = Vec::new();
        let raw_text = " ";

        let result = process_syllable_text(raw_text, &mut syllables);

        // 确保不会崩溃
        assert_eq!(result, None);
        assert!(syllables.is_empty());
    }

    // 对一个已经有尾随空格的音节，再次添加空格
    #[test]
    fn test_space_is_idempotent() {
        let mut syllables = vec![LyricSyllable {
            text: "previous".to_string(),
            ends_with_space: true,
            ..Default::default()
        }];
        let raw_text = " ";

        let result = process_syllable_text(raw_text, &mut syllables);

        assert_eq!(result, None);
        assert_eq!(syllables.len(), 1);
        assert!(syllables[0].ends_with_space, "尾随空格标志应保持不变");
    }
}
