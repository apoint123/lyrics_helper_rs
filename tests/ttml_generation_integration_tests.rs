use lyrics_helper_rs::converter::{
    generators::ttml_generator::generate_ttml,
    processors::metadata_processor::MetadataStore,
    types::{
        AgentStore, AnnotatedTrack, CanonicalMetadataKey, ContentType, LyricLineBuilder,
        LyricSyllableBuilder, LyricTrack, TrackMetadataKey, TtmlGenerationOptionsBuilder,
        TtmlTimingMode, Word,
    },
};

use std::collections::HashMap;

#[test]
fn test_generate_line_timed_snapshot() {
    let lines = vec![
        LyricLineBuilder::default()
            .start_ms(1000)
            .end_ms(5000)
            .track(AnnotatedTrack {
                content_type: ContentType::Main,
                content: LyricTrack {
                    words: vec![Word {
                        syllables: vec![
                            LyricSyllableBuilder::default()
                                .text("这是一行歌词")
                                .build()
                                .unwrap(),
                        ],
                        ..Default::default()
                    }],
                    ..Default::default()
                },
                ..Default::default()
            })
            .build()
            .unwrap(),
        LyricLineBuilder::default()
            .start_ms(6000)
            .end_ms(10000)
            .track(AnnotatedTrack {
                content_type: ContentType::Main,
                content: LyricTrack {
                    words: vec![Word {
                        syllables: vec![
                            LyricSyllableBuilder::default()
                                .text("这是第二行歌词")
                                .build()
                                .unwrap(),
                        ],
                        ..Default::default()
                    }],
                    ..Default::default()
                },
                ..Default::default()
            })
            .build()
            .unwrap(),
    ];
    let mut metadata = MetadataStore::new();
    metadata.set_single(&CanonicalMetadataKey::Title.to_string(), "逐行歌曲");
    metadata.set_single(&CanonicalMetadataKey::Artist.to_string(), "测试艺术家");

    let options = TtmlGenerationOptionsBuilder::default()
        .timing_mode(TtmlTimingMode::Line)
        .format(true)
        .build()
        .unwrap();

    let agent_store = AgentStore::from_metadata_store(&metadata);
    let ttml_output = generate_ttml(&lines, &metadata, &agent_store, &options).unwrap();

    insta::assert_snapshot!(ttml_output);
}

#[test]
fn test_generate_word_timed_with_agents_snapshot() {
    let lines = vec![
        LyricLineBuilder::default()
            .agent(Some("v1".to_string()))
            .start_ms(5000)
            .end_ms(8200)
            .track(AnnotatedTrack {
                content_type: ContentType::Main,
                content: LyricTrack {
                    words: vec![Word {
                        syllables: vec![
                            LyricSyllableBuilder::default()
                                .text("I")
                                .start_ms(5000)
                                .end_ms(5500)
                                .ends_with_space(true)
                                .build()
                                .unwrap(),
                            LyricSyllableBuilder::default()
                                .text("sing")
                                .start_ms(5600)
                                .end_ms(6200)
                                .build()
                                .unwrap(),
                        ],
                        ..Default::default()
                    }],
                    ..Default::default()
                },
                ..Default::default()
            })
            .build()
            .unwrap(),
        LyricLineBuilder::default()
            .agent(Some("v1000".to_string()))
            .start_ms(9000)
            .end_ms(12000)
            .track(AnnotatedTrack {
                content_type: ContentType::Main,
                content: LyricTrack {
                    words: vec![Word {
                        syllables: vec![
                            LyricSyllableBuilder::default()
                                .text("We")
                                .start_ms(9000)
                                .end_ms(9500)
                                .ends_with_space(true)
                                .build()
                                .unwrap(),
                            LyricSyllableBuilder::default()
                                .text("sing")
                                .start_ms(9600)
                                .end_ms(10200)
                                .build()
                                .unwrap(),
                        ],
                        ..Default::default()
                    }],
                    ..Default::default()
                },
                ..Default::default()
            })
            .build()
            .unwrap(),
    ];

    let mut metadata = MetadataStore::new();
    metadata
        .add(&CanonicalMetadataKey::Songwriter.to_string(), "作曲家1号")
        .unwrap();

    metadata.add("agent", "v1=演唱者1号").unwrap();
    metadata.add("agent", "v1000=合唱").unwrap();

    let options = TtmlGenerationOptionsBuilder::default()
        .timing_mode(TtmlTimingMode::Word)
        .use_apple_format_rules(true)
        .format(true)
        .build()
        .unwrap();

    let agent_store = AgentStore::from_metadata_store(&metadata);
    let ttml_output = generate_ttml(&lines, &metadata, &agent_store, &options).unwrap();

    insta::assert_snapshot!(ttml_output);
}

#[test]
fn test_auto_word_splitting_snapshot() {
    let lines = vec![
        LyricLineBuilder::default()
            .start_ms(1000)
            .end_ms(5000)
            .track(AnnotatedTrack {
                content_type: ContentType::Main,
                content: LyricTrack {
                    words: vec![Word {
                        syllables: vec![
                            LyricSyllableBuilder::default()
                                .text("Split this,你好世界")
                                .start_ms(1000)
                                .end_ms(5000)
                                .build()
                                .unwrap(),
                        ],
                        ..Default::default()
                    }],
                    ..Default::default()
                },
                ..Default::default()
            })
            .build()
            .unwrap(),
    ];

    let options = TtmlGenerationOptionsBuilder::default()
        .timing_mode(TtmlTimingMode::Word)
        .auto_word_splitting(true)
        .format(true)
        .punctuation_weight(0.1)
        .build()
        .unwrap();

    let metadata = Default::default();
    let agent_store = AgentStore::from_metadata_store(&metadata);
    let ttml_output = generate_ttml(&lines, &metadata, &agent_store, &options).unwrap();

    insta::assert_snapshot!(ttml_output);
}

#[test]
fn test_generate_timed_romanization_snapshot() {
    let mut main_track_metadata = HashMap::new();
    main_track_metadata.insert(
        TrackMetadataKey::Custom("itunes_key".to_string()),
        "L1".to_string(),
    );
    let mut romanization_track_metadata = HashMap::new();
    romanization_track_metadata.insert(TrackMetadataKey::Language, "ja-Latn".to_string());

    let main_syllable = LyricSyllableBuilder::default()
        .text("朝も")
        .start_ms(1000)
        .end_ms(2000)
        .build()
        .unwrap();

    let roma_syllables = vec![
        LyricSyllableBuilder::default()
            .text("Asa")
            .start_ms(1000)
            .end_ms(1500)
            .build()
            .unwrap(),
        LyricSyllableBuilder::default()
            .text("mo")
            .start_ms(1600)
            .end_ms(2000)
            .build()
            .unwrap(),
    ];

    let lines = vec![
        LyricLineBuilder::default()
            .start_ms(1000)
            .end_ms(2000)
            .track(AnnotatedTrack {
                content_type: ContentType::Main,
                content: LyricTrack {
                    words: vec![Word {
                        syllables: vec![main_syllable],
                        ..Default::default()
                    }],
                    ..Default::default()
                },
                romanizations: vec![LyricTrack {
                    words: vec![Word {
                        syllables: roma_syllables,
                        ..Default::default()
                    }],
                    metadata: romanization_track_metadata,
                }],
                ..Default::default()
            })
            .build()
            .unwrap(),
    ];

    let options = TtmlGenerationOptionsBuilder::default()
        .timing_mode(TtmlTimingMode::Word)
        .use_apple_format_rules(true)
        .format(true)
        .build()
        .unwrap();

    let metadata = Default::default();
    let agent_store = AgentStore::from_metadata_store(&metadata);
    let ttml_output = generate_ttml(&lines, &metadata, &agent_store, &options).unwrap();

    insta::assert_snapshot!(ttml_output);
}

#[test]
fn test_generate_timed_translation_snapshot() {
    let mut main_track_metadata = HashMap::new();
    main_track_metadata.insert(
        TrackMetadataKey::Custom("itunes_key".to_string()),
        "L1".to_string(),
    );
    let mut translation_track_metadata = HashMap::new();
    translation_track_metadata.insert(TrackMetadataKey::Language, "zh-Hans".to_string());

    let main_syllable = LyricSyllableBuilder::default()
        .text("鐘聲響起歸家")
        .start_ms(1000)
        .end_ms(2000)
        .build()
        .unwrap();

    let trans_syllables = vec![
        LyricSyllableBuilder::default()
            .text("钟声响起")
            .start_ms(1000)
            .end_ms(1500)
            .build()
            .unwrap(),
        LyricSyllableBuilder::default()
            .text("归家")
            .start_ms(1500)
            .end_ms(2000)
            .build()
            .unwrap(),
    ];

    let lines = vec![
        LyricLineBuilder::default()
            .start_ms(1000)
            .end_ms(2000)
            .track(AnnotatedTrack {
                content_type: ContentType::Main,
                content: LyricTrack {
                    words: vec![Word {
                        syllables: vec![main_syllable],
                        ..Default::default()
                    }],
                    ..Default::default()
                },
                translations: vec![LyricTrack {
                    words: vec![Word {
                        syllables: trans_syllables,
                        ..Default::default()
                    }],
                    metadata: translation_track_metadata,
                }],
                ..Default::default()
            })
            .build()
            .unwrap(),
    ];

    let options = TtmlGenerationOptionsBuilder::default()
        .timing_mode(TtmlTimingMode::Word)
        .use_apple_format_rules(true)
        .format(true)
        .build()
        .unwrap();

    let metadata = Default::default();
    let agent_store = AgentStore::from_metadata_store(&metadata);
    let ttml_output = generate_ttml(&lines, &metadata, &agent_store, &options).unwrap();

    insta::assert_snapshot!(ttml_output);
}
