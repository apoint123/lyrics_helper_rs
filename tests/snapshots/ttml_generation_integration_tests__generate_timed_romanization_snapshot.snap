---
source: tests/ttml_generation_integration_tests.rs
expression: ttml_output
---
<tt xmlns="http://www.w3.org/ns/ttml"
    xmlns:itunes="http://music.apple.com/lyric-ttml-internal"
    xmlns:ttm="http://www.w3.org/ns/ttml#metadata"
    itunes:timing="Word">
  <head>
    <metadata>
      <ttm:agent type="person" xml:id="v1"/>
      <iTunesMetadata xmlns="http://music.apple.com/lyric-ttml-internal">
        <transliterations>
          <transliteration xml:lang="ja-Latn">
            <text for="L1">
              <span begin="1.000" end="1.500">Asa</span>
              <span begin="1.600" end="2.000">mo</span>
            </text>
          </transliteration>
        </transliterations>
      </iTunesMetadata>
    </metadata>
  </head>
  <body dur="2.000">
    <div begin="1.000" end="2.000">
      <p begin="1.000" end="2.000" itunes:key="L1" ttm:agent="v1">
        <span begin="1.000" end="2.000">朝も</span>
      </p>
    </div>
  </body>
</tt>
